### ISSUE TITLE
have trouble about resume from ckpt

### ISSUE THREAD

**Issue by SouppuoS** [mpariente/asteroid#322]
I try to train DPRNN and DPTNet on wsj0-3mix from a checkpoint. and val_loss look like this (blue line is train from a ckpt). 
I think it should be a curve much smoother.
![snap](https://user-images.githubusercontent.com/11723579/98638931-b11b7300-2364-11eb-93f6-dcc57058616e.png)

Scripts were copied from wham folder. to resume from ckpt I add some code in train.py like this:
```
state_dict = torch.load(os.path.join(exp_dir, ckpt_name))   
system.load_state_dict(state_dict['state_dict'])
...
trainer = pl.Trainer(..., resume_from_checkpoint=os.path.join(exp_dir, ckpt_name))
```
And to train DPRNN & DPTNet on wsj0-3mix, I replace wham dataset & dataloader code in train.py to this:
```
train_loader, val_loader = make_dataloaders(n_src=3, **conf["data"], **conf["training"])
```
no change in conf.yml.
I also tried deepclustering on wsj0-3mix, dprnn & dptnet on wham and my own dataset, did not find this problem.

what did I miss? Need help plz.

**Comment by mpariente** [comment]
Thanks for the detailed issue. 

First guess: this might be related to the optimizer and scheduler. I don't know if the optimizer state is resumed from the checkpoint. Also don't know if the learning rate was halved, does it start from half again? 

I didn't encounter this problem so I'm not sure what's the solution. Can you check if the optimizer's state is loaded? (or even present in the checkpoint?)

**Comment by jonashaag** [comment]
I am seeing this regularly when interrupting and then continuing a training, with all sorts of optimizers. But what I haven't seen so far is that before the interruption training essentially stagnated when compared to what happens to the training after continuation.

Can you post a validation loss plot with less smoothing? Maybe it has something to do with how Tensorboard smoothes.

**Comment by SouppuoS** [comment]
hi, mpariente 
I load the ckpt (generated from DPTNet) and there is nothing in `lr_schedulers`. I dont know if it is right.
![snap2](https://user-images.githubusercontent.com/11723579/98671223-e341ca80-238e-11eb-82df-7c9b6794bac2.jpg)
and after `system.load_state_dict()`, system.scheduler seems did not change, as is shown below. I think the epoch should be 83 (83 can be found in ckpt's state_dict['epoch']) instead of 6. 
![snap1](https://user-images.githubusercontent.com/11723579/98670009-04092080-238d-11eb-9fc6-424c39aa9a5a.png)

hi, jonashaag
I plot the val_loss with 0 smooth. The gap is so huge (-6.5 vs -5, about 1.5dB).
![loss2](https://user-images.githubusercontent.com/11723579/98670612-e7211d00-238d-11eb-8c91-52eb387f303b.png)

**Comment by jonashaag** [comment]
Do you also have a learning rate plot in Tensorboard? Can use you it to double check if it's really taking the epoch 6 LR? (Which, as Manuel guessed, a likely reason for the peak)

**Comment by mpariente** [comment]
But your final performance is better! Maybe you've discovered a great learning technique :star_struck: 

Some weird things are happening, that's for sure.. 
Do you have a plot of the training loss as well? 
What happens with DPRNN? (where the scheduler is much simpler?)

> Do you also have a learning rate plot in Tensorboard?

I think we don't log that but maybe we should. 

BTW, what do you use to visualize objects like that? It looks cool !

**Comment by SouppuoS** [comment]
I think it is at epoch 6. I resume from ckpt epoch 83, plot lr (which is got by `self.scheduler[0]['scheduler']._get_lr()` and `self.optimizer.param_groups[0]['lr]` in system.py), and found lr is increasing. I notice it is a feature of DPTNet, lr will increasing when first 4000 steps.

**Comment by mpariente** [comment]
@popcornell is the specialist about this here but yeah, this is called the warmup. 
You can use the `plot` and `as_tensor` methods to see the learning rate schedule of `DPTNetScheduler`.

**Comment by SouppuoS** [comment]
I mean it should not happened when it is already epoch 86 (maybe 100000 step already). I think your guess is right, optimizer state is not resumed from the checkpoint. I'll also check this in `DPTNetScheduler`
curve of train_loss is similar with val_loss, a gap between ckpt.
I debug it in vscode, so i can know the detail of the objects. : )

**Comment by jonashaag** [comment]
Could it be that this is bug in the scheduler? In the PyTorch scheduler base class there is some state/serialization code that we are not implementing: https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#LambdaLR

**Comment by mpariente** [comment]
I think that's related only to the lambdas in the scheduler. 
But actually, we keep the optimizer in the keys (and the state_dict) and that's probably wrong. 
```python
# In torch
def state_dict(self):
    return {key: value for key, value in self.__dict__.items() if key != 'optimizer'}
# In asteroid
def state_dict(self):
    return {key: value for key, value in self.__dict__.items()}
```

I don't know what happens to the schedulers during checkpointing in lightning, I'd have to check. 
@SouppuoS, would you like to have a look?

**Comment by popcornell** [comment]
My guess is that there could be some issues regarding saving of schedulers because we use a `dict`  to enforce the scheduler to be called at every step rather than at every epoch (torch.optim schedulers are designed to work only at end of epoch and lightning up till 0.9.0 at least did not support scheduling on each batch).
If my guess is true there will not be any problem for schedulers which are not passed as a dict (e.g. standard torch schedulers). 
https://github.com/mpariente/asteroid/blob/92e74ebdf9a9d513554b42169f3ba789bd12a0f5/asteroid/engine/system.py#L168 

What we can do ? 
Ligthning apparently do not like dicts. We can include the interval as an attribute for the custom scheduler class so we don't need a `dict`. This should be fairly straightforward

**Comment by mpariente** [comment]
My take: leave it like this for now, even if it is bugged and work on solving it when we switch to pytorch-lightning 1.0, WDYT?

**Comment by popcornell** [comment]
UPDATE: the scheduler is saved and loaded correctly when it is not a `dict` (had to hack and make it a child of torch.optim.lr_schedulers._LRScheduler ). 
I think it must be a lightning issue as we are basically doing nothing special, schedulers as dicts are supported in lightning:

```python
        lr_schedulers = []
        default_config = {'interval': 'epoch',  # default every epoch
                          'frequency': 1,  # default every epoch/batch
                          'reduce_on_plateau': False,  # most often not ReduceLROnPlateau scheduler
                          'monitor': 'val_loss'}  # default value to monitor for ReduceLROnPlateau
        for scheduler in schedulers:
            if isinstance(scheduler, dict):
                if 'scheduler' not in scheduler:
                    raise ValueError('Lr scheduler should have key `scheduler`',
                                     ' with item being a lr scheduler')
                scheduler['reduce_on_plateau'] = isinstance(
                    scheduler['scheduler'], optim.lr_scheduler.ReduceLROnPlateau)

                lr_schedulers.append({**default_config, **scheduler})
```

**Comment by mpariente** [comment]
Cool ! The other question is: is the scheduler's resumed when using `resume_from_checkpoint`?

What about this difference: 
```
# In torch
def state_dict(self):
    return {key: value for key, value in self.__dict__.items() if key != 'optimizer'}
# In asteroid
def state_dict(self):
    return {key: value for key, value in self.__dict__.items()}
```
I'm not sure the optimizer should be in the state_dict..

**Comment by popcornell** [comment]
I have a pull request which should solve this issue. Need feedback
